File: android/app/src/main/java/ai/oriveo/community/core/provider/transport/EndpointResolver.kt
package ai.oriveo.community.core.provider.transport

import ai.oriveo.community.core.model.Provider
import ai.oriveo.community.core.model.ProviderKind
import java.net.URI

/**
 * Endpoint URL resolver.
 *
 * Three levels of priority:
 *   1. the [Provider.baseUrlText] the user typed into the provider settings (a self-hosted
 *      proxy, Azure OpenAI, a regional gateway, and so on)
 *   2. the [ProviderTransportDefinition.baseUrl] and endpoints from the published catalog
 *   3. the built-in fallback table, as a safety net
 *
 * A Relay provider always requires an explicit base URL and gets no transport from the catalog.
 */
object EndpointResolver {

    /** DashScope's OpenAI-compatible endpoint prefix, used by the Qwen base-normalisation case. */
    private const val QWEN_COMPATIBLE_PREFIX = "/compatible-mode/v1"

    /** Endpoint kinds, aligned one-to-one with the fields of [TransportEndpoints]. */
    enum class EndpointKind {
        CHAT,
        RESPONSES,
        IMAGES,
        EMBEDDINGS,
        FILES,
    }

    /**
     * Resolves the full URL for this provider and endpoint kind.
     *
     * @param provider the current provider instance, including any user-supplied baseUrlText
     * @param kind which endpoint is being addressed
     * @param metadata the provider transport from the published catalog, when available
     * @return a full URL including scheme, host and path; falls back to the built-in table
     */
    fun resolveEndpoint(
        provider: Provider,
        kind: EndpointKind,
        metadata: ProviderTransportDefinition? = null,
    ): String {
        val userBase = provider.baseUrlText?.trim()?.takeIf { it.isNotEmpty() }
        val userPath = endpointPathFor(metadata?.endpoints, kind)

        // Priority, merged: user baseUrl + user endpoint > user baseUrl + catalog endpoint
        //                   > catalog baseUrl + catalog endpoint
        //                   > built-in fallback baseUrl + built-in fallback path
        val metadataBase = metadata?.baseUrl?.trim()?.takeIf { it.isNotEmpty() }
        val resolvedBase = userBase
            ?: metadataBase?.takeIf { isAllowedOfficialMetadataBase(provider.kind, it) }
            ?: fallbackBaseUrl(provider.kind)

        val resolvedPath = userPath
            ?: fallbackPath(provider.kind, kind)
            ?: ""

        // A version segment already present in the base has to come off before the path is
        // appended, or `/v1` + `/v1/chat/completions` produces a 404. See below.
        val normalizedBase = normalizeBaseForEndpoint(resolvedBase, provider.kind, resolvedPath)

        return joinUrl(normalizedBase, resolvedPath)
    }

    /**
     * De-duplicates the version segment between the base and the endpoint path.
     *
     * The catalog contract is "baseUrl stops at the origin, endpoints.* carries the full version
     * path" (deepseek is `https://api.deepseek.com` plus `/v1/chat/completions`), while an
     * official provider's [Provider.baseUrlText] holds [ProviderKind.defaultBaseUrl]
     * (`api.deepseek.com/v1`, version segment included). Concatenating the two naively yields
     * `/v1/v1/chat/completions`, and the upstream answers 404 with an empty body.
     *
     * The base is only trimmed back to its origin when the path really does cover the prefix the
     * base already carries; everything else is returned unchanged, so the version-less fallback
     * shape (base = origin plus path = /v1/...) behaves exactly as before.
     */
    private fun normalizeBaseForEndpoint(
        base: String,
        providerKind: ProviderKind,
        endpoint: String,
    ): String {
        if (endpoint.isEmpty()) return base
        // An absolute endpoint URL has its own joinUrl semantics; the base takes no part.
        if (endpoint.startsWith("http://", ignoreCase = true) ||
            endpoint.startsWith("https://", ignoreCase = true)
        ) {
            return base
        }

        val uri = runCatching { URI(ensureScheme(base).trimEnd('/')) }.getOrNull() ?: return base
        val scheme = uri.scheme ?: return base
        val authority = uri.authority ?: return base
        val basePath = uri.path.orEmpty().trimEnd('/')
        if (basePath.isEmpty()) return base

        val endpointPath = pathnameForPath(endpoint)
        if (endpointPath.isEmpty()) return base

        val origin = "$scheme://$authority"

        // Qwen special case: chat goes to /compatible-mode/v1/... while images go to the native
        // /api/v1/... path. If the user's base stops on the compatible prefix, both paths have
        // to strip it first - images do not overlap the base at all, so the general rule below
        // never fires for them.
        if (providerKind == ProviderKind.Qwen && basePath.endsWith(QWEN_COMPATIBLE_PREFIX)) {
            return origin + basePath.removeSuffix(QWEN_COMPATIBLE_PREFIX)
        }

        if (endpointPath == basePath || endpointPath.startsWith("$basePath/")) {
            return origin
        }

        return base
    }

    /** The bare path part of an endpoint: query and fragment dropped, trailing slash removed,
     *  leading slash added. */
    private fun pathnameForPath(path: String): String {
        val trimmed = path.substringBefore('#').substringBefore('?').trimEnd('/')
        if (trimmed.isEmpty()) return ""
        return if (trimmed.startsWith("/")) trimmed else "/$trimmed"
    }

    /** For services that assemble their own URLs: returns the base URL only, without appending
     *  an endpoint path. */
    fun resolveBaseUrl(
        provider: Provider,
        metadata: ProviderTransportDefinition? = null,
    ): String {
        val userBase = provider.baseUrlText?.trim()?.takeIf { it.isNotEmpty() }
        val metadataBase = metadata?.baseUrl?.trim()?.takeIf { it.isNotEmpty() }
        val base = userBase
            ?: metadataBase?.takeIf { isAllowedOfficialMetadataBase(provider.kind, it) }
            ?: fallbackBaseUrl(provider.kind)
        return ensureScheme(base).trimEnd('/')
    }

    private fun isAllowedOfficialMetadataBase(kind: ProviderKind, rawBase: String): Boolean {
        if (kind == ProviderKind.Relay) return true
        val normalized = ensureScheme(rawBase).trimEnd('/')
        val uri = runCatching { URI(normalized) }.getOrNull()
        val host = uri?.host?.lowercase()
        val allowedDomains = OFFICIAL_METADATA_BASE_ETLD1[kind].orEmpty()
        return uri?.scheme == "https" &&
            host != null &&
            allowedDomains.any { host == it || host.endsWith(".$it") }
    }

    private fun endpointPathFor(endpoints: TransportEndpoints?, kind: EndpointKind): String? {
        endpoints ?: return null
        return when (kind) {
            EndpointKind.CHAT -> endpoints.chat
            EndpointKind.RESPONSES -> endpoints.responses
            EndpointKind.IMAGES -> endpoints.images
            EndpointKind.EMBEDDINGS -> endpoints.embeddings
            EndpointKind.FILES -> endpoints.files
        }?.takeIf { it.isNotBlank() }
    }

    /**
     * The built-in fallback base URL, used only when catalog metadata is entirely unavailable.
     *
     * **Delegates to [ProviderKind.defaultBaseUrl]**, which is the single source of truth for
     * official endpoints and is also what provider auto-fill uses. This used to be a second copy
     * of that table, and the copy had dropped four path prefixes (openRouter was missing `/api`,
     * groq `/openai`, fireworks `/inference`, zhipu `/api/paas`), so every fallback landed on a
     * 404.
     *
     * It takes the origin rather than the whole string, matching the catalog contract of
     * "baseUrl to the origin, endpoints carry the full version path"; the path half is supplied
     * by [fallbackPath] from the same table (see [fallbackVersionPrefix]).
     */
    private fun fallbackBaseUrl(kind: ProviderKind): String = when (kind) {
        // Relay requires an explicit base URL, so there is nothing to fall back to.
        ProviderKind.Relay -> ""
        else -> kind.defaultBaseUrl?.let(::originOf).orEmpty()
    }

    /** The `scheme://authority` of a URL; returned unchanged when it cannot be parsed (with a
     *  scheme added and any trailing slash removed). */
    private fun originOf(baseUrl: String): String {
        val normalized = ensureScheme(baseUrl).trimEnd('/')
        val uri = runCatching { URI(normalized) }.getOrNull() ?: return normalized
        val scheme = uri.scheme ?: return normalized
        val authority = uri.authority ?: return normalized
        return "$scheme://$authority"
    }

    /**
     * The version-segment prefix for OpenAI-compatible fallback paths, taken from the path part
     * of [ProviderKind.defaultBaseUrl].
     *
     * That way base (origin) plus path (including the prefix) reconstructs the real official
     * endpoint, and it still lines up in the common case where the user's baseUrl happens to be
     * exactly defaultBaseUrl ([normalizeBaseForEndpoint] trims the duplicated segment). When the
     * table has no path at all - Relay, or DashScope's native origin - it follows the OpenAI
     * convention of `/v1`.
     */
    private fun fallbackVersionPrefix(kind: ProviderKind): String {
        val base = kind.defaultBaseUrl ?: return "/v1"
        val path = runCatching { URI(ensureScheme(base).trimEnd('/')).path }.getOrNull().orEmpty()
        return path.ifEmpty { "/v1" }
    }

    private val OFFICIAL_METADATA_BASE_ETLD1: Map<ProviderKind, Set<String>> = mapOf(
        ProviderKind.OpenAI to setOf("openai.com"),
        ProviderKind.Anthropic to setOf("anthropic.com"),
        ProviderKind.Gemini to setOf("googleapis.com"),
        ProviderKind.DeepSeek to setOf("deepseek.com"),
        ProviderKind.Grok to setOf("x.ai"),
        ProviderKind.OpenRouter to setOf("openrouter.ai"),
        ProviderKind.Groq to setOf("groq.com"),
        ProviderKind.Together to setOf("together.xyz"),
        ProviderKind.Fireworks to setOf("fireworks.ai"),
        ProviderKind.MiniMax to setOf("minimax.io"),
        ProviderKind.Zhipu to setOf("bigmodel.cn"),
        ProviderKind.Qwen to setOf("aliyuncs.com"),
        ProviderKind.Moonshot to setOf("moonshot.ai", "moonshot.cn"),
        ProviderKind.Mistral to setOf("mistral.ai"),
        ProviderKind.SiliconFlow to setOf("siliconflow.cn", "siliconflow.com"),
    )

    /**
     * The built-in fallback endpoint paths, aligned with the catalog's default transport table.
     */
    private fun fallbackPath(provider: ProviderKind, kind: EndpointKind): String? {
        return when (provider) {
            ProviderKind.OpenAI -> when (kind) {
                EndpointKind.CHAT -> "/v1/chat/completions"
                EndpointKind.RESPONSES -> "/v1/responses"
                EndpointKind.IMAGES -> "/v1/images/generations"
                EndpointKind.FILES -> "/v1/files"
                EndpointKind.EMBEDDINGS -> "/v1/embeddings"
            }
            ProviderKind.Anthropic -> when (kind) {
                EndpointKind.CHAT -> "/v1/messages"
                EndpointKind.FILES -> "/v1/files"
                else -> null
            }
            ProviderKind.Gemini -> when (kind) {
                EndpointKind.CHAT -> "/v1beta"
                EndpointKind.IMAGES -> "/v1beta"
                else -> null
            }
            ProviderKind.Qwen -> when (kind) {
                // Chat uses the OpenAI-compatible endpoint: DashScope's native text-generation
                // path is retired, and qwen3.x answers it with an in-stream error that gets
                // swallowed into an empty response.
                EndpointKind.CHAT -> "/compatible-mode/v1/chat/completions"
                EndpointKind.IMAGES -> "/api/v1/services/aigc/multimodal-generation/generation"
                else -> null
            }
            ProviderKind.MiniMax -> when (kind) {
                // Text uses the OpenAI-compatible path; image models use MiniMax's own endpoint.
                // The OpenAI-shaped /v1/images/generations does not exist on MiniMax, so falling
                // into the generic bucket is a guaranteed 404.
                EndpointKind.CHAT -> "/v1/chat/completions"
                EndpointKind.IMAGES -> "/v1/image_generation"
                else -> null
            }
            ProviderKind.Grok,
            ProviderKind.DeepSeek,
            ProviderKind.OpenRouter,
            ProviderKind.Groq,
            ProviderKind.Together,
            ProviderKind.Fireworks,
            ProviderKind.Zhipu,
            ProviderKind.Moonshot,
            ProviderKind.Mistral,
            ProviderKind.SiliconFlow,
            ProviderKind.Relay -> fallbackVersionPrefix(provider).let { prefix ->
                when (kind) {
                    EndpointKind.CHAT -> "$prefix/chat/completions"
                    EndpointKind.RESPONSES -> "$prefix/responses"
                    EndpointKind.IMAGES -> "$prefix/images/generations"
                    EndpointKind.FILES -> "$prefix/files"
                    EndpointKind.EMBEDDINGS -> "$prefix/embeddings"
                }
            }
        }
    }

    /** Joins a URL, accepting either a host-only or a host-plus-path base. */
    private fun joinUrl(base: String, path: String): String {
        val scheme = ensureScheme(base).trimEnd('/')
        if (path.isEmpty()) return scheme
        val normalizedPath = if (path.startsWith("/")) path else "/$path"
        return "$scheme$normalizedPath"
    }

    private fun ensureScheme(url: String): String {
        val trimmed = url.trim()
        return if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
            trimmed
        } else {
            "https://$trimmed"
        }
    }
}

File: ios/Oriveo/Oriveo/Core/Providers/Transport/EndpointResolver.swift
import Foundation

enum EndpointResolver {

    struct ResolvedEndpoint: Equatable, Sendable {
        let baseURL: String
        let path: String
        /// `baseURL` joined with `path`. `joinURL` normalizes the seam between them, so a
        /// trailing slash on the base or a missing leading slash on the path cannot produce
        /// a doubled or glued path segment.
        let url: URL
    }

    struct EndpointResolutionError: Error, Sendable {
        let providerKind: ProviderKind
        let endpoint: EndpointKind
        let reason: String
    }

    /// - Parameters:
    ///   - provider: the provider whose `baseURLText`, when set, overrides both the metadata
    ///     base URL and the built-in fallback.
    /// - Returns: the base URL, path and joined URL actually used for the request.
    /// - Throws: `EndpointResolutionError` when base and path cannot form a valid URL.
    static func resolve(
        provider: Provider,
        kind: EndpointKind,
        metadataTransport: MetadataClient.ProviderTransportDefinition?
    ) throws -> ResolvedEndpoint {
        try resolve(
            providerKind: provider.kind,
            userBaseURL: provider.baseURLText,
            kind: kind,
            metadataTransport: metadataTransport
        )
    }

    /// Resolves an endpoint by precedence: user-supplied base URL, then the base URL from
    /// server metadata, then the built-in fallback.
    ///
    /// A metadata base URL is only trusted for an official provider when its host is on that
    /// provider's allowlist, so a stale or tampered metadata payload cannot silently redirect
    /// traffic that carries the user's API key.
    static func resolve(
        providerKind: ProviderKind,
        userBaseURL: String? = nil,
        kind: EndpointKind,
        metadataTransport: MetadataClient.ProviderTransportDefinition?
    ) throws -> ResolvedEndpoint {
        let userBase = sanitize(userBaseURL)
        let rawMetaBase = sanitize(metadataTransport?.baseUrl)
        let metaBase = validatedOfficialMetadataBaseURL(rawMetaBase, providerKind: providerKind)
        let metaPath = sanitize(endpointPath(in: metadataTransport?.endpoints, kind: kind))

        let fallbackBase = fallbackBaseURL(for: providerKind)
        let fallbackPath = fallbackEndpointPath(providerKind, kind: kind)

        let chosenBase: String
        let chosenPath: String
        // Metadata paths are absolute from the origin and carry their own version prefix
        // (`/v1/chat/completions`), while fallback paths are relative to a base URL that
        // already contains that prefix. Remember which one supplied the path so the base can
        // be trimmed back to the origin first; otherwise the prefix ends up in the URL twice.
        let pathFromMetadata: Bool
        if let userBase {
            chosenBase = userBase
            chosenPath = metaPath ?? fallbackPath
            pathFromMetadata = metaPath != nil
        } else if let metaBase, let metaPath {
            chosenBase = metaBase
            chosenPath = metaPath
            pathFromMetadata = true
        } else if let metaBase {
            chosenBase = metaBase
            chosenPath = fallbackPath
            pathFromMetadata = false
        } else {
            chosenBase = fallbackBase
            chosenPath = fallbackPath
            pathFromMetadata = false
        }

        let qwenImagesNeedsNativeOrigin = providerKind == .qwen && kind == .images
        let effectiveBase = pathFromMetadata || qwenImagesNeedsNativeOrigin
            ? normalizeBaseForEndpoint(base: chosenBase, providerKind: providerKind, endpointPath: chosenPath)
            : chosenBase

        guard let url = joinURL(base: effectiveBase, path: chosenPath) else {
            throw EndpointResolutionError(
                providerKind: providerKind,
                endpoint: kind,
                reason: "Failed to join baseURL=\(effectiveBase) path=\(chosenPath)"
            )
        }
        return ResolvedEndpoint(baseURL: effectiveBase, path: chosenPath, url: url)
    }

    static func resolveBaseURL(
        provider: Provider,
        metadataTransport: MetadataClient.ProviderTransportDefinition?
    ) -> String {
        if let userBase = sanitize(provider.baseURLText) { return userBase }
        if let metaBase = validatedOfficialMetadataBaseURL(
            sanitize(metadataTransport?.baseUrl),
            providerKind: provider.kind
        ) {
            return metaBase
        }
        return fallbackBaseURL(for: provider.kind)
    }

    static func officialMetadataBaseURL(
        providerKind: ProviderKind,
        metadataTransport: MetadataClient.ProviderTransportDefinition?
    ) -> String? {
        validatedOfficialMetadataBaseURL(sanitize(metadataTransport?.baseUrl), providerKind: providerKind)
    }

    // MARK: - URL assembly

    static func joinURL(base: String, path: String) -> URL? {
        let trimmedBase = base.hasSuffix("/") ? String(base.dropLast()) : base
        let trimmedPath = path.hasPrefix("/") ? path : "/" + path
        return URL(string: trimmedBase + trimmedPath)
    }

    /// Trims a base URL back to its origin so a metadata endpoint path can be appended cleanly.
    ///
    /// Metadata pairs `transport.baseUrl` with `endpoints.*` paths that are absolute from the
    /// origin (`https://api.deepseek.com` + `/v1/chat/completions`), so a base that already
    /// carries the version prefix would double it.
    ///
    /// Only call this when the path came from metadata. Fallback paths are relative to the
    /// base — qwen's fallback base ends in `/compatible-mode/v1` and its chat path is just
    /// `/chat/completions` — and trimming would drop a segment the path never restates.
    private static func normalizeBaseForEndpoint(
        base: String,
        providerKind: ProviderKind,
        endpointPath: String
    ) -> String {
        let lowered = endpointPath.lowercased()
        guard !lowered.isEmpty, !lowered.hasPrefix("http://"), !lowered.hasPrefix("https://") else {
            return base
        }
        guard var components = URLComponents(string: base),
              components.scheme != nil,
              components.host != nil else {
            return base
        }

        let path = normalizedPathname(endpointPath)
        guard !path.isEmpty else { return base }

        let basePath = trimTrailingSlashes(components.path)
        guard !basePath.isEmpty else { return base }

        let stripToPrefix: String?
        let qwenCompatibleSuffix = "/compatible-mode/v1"
        if providerKind == .qwen, basePath.hasSuffix(qwenCompatibleSuffix) {
            // Qwen's native DashScope paths (`/api/v1/services/...`) do not live under the
            // OpenAI-compatible prefix, so drop `/compatible-mode/v1` and rejoin from the origin.
            stripToPrefix = String(basePath.dropLast(qwenCompatibleSuffix.count))
        } else if path == basePath || path.hasPrefix(basePath + "/") {
            stripToPrefix = ""
        } else {
            stripToPrefix = nil
        }
        guard let stripToPrefix else { return base }

        components.path = stripToPrefix
        components.query = nil
        components.fragment = nil
        return trimTrailingSlashes(components.string ?? base)
    }

    private static func normalizedPathname(_ path: String) -> String {
        var value = path
        if let cut = value.firstIndex(where: { $0 == "?" || $0 == "#" }) {
            value = String(value[value.startIndex..<cut])
        }
        if !value.hasPrefix("/") { value = "/" + value }
        return trimTrailingSlashes(value)
    }

    private static func trimTrailingSlashes(_ value: String) -> String {
        var value = value
        while value.hasSuffix("/") { value.removeLast() }
        return value
    }

    private static func sanitize(_ value: String?) -> String? {
        guard let v = value?.trimmingCharacters(in: .whitespacesAndNewlines), !v.isEmpty else {
            return nil
        }
        return v
    }

    private static func endpointPath(
        in endpoints: MetadataClient.TransportEndpoints?,
        kind: EndpointKind
    ) -> String? {
        guard let endpoints else { return nil }
        switch kind {
        case .chat: return endpoints.chat
        case .responses: return endpoints.responses
        case .images: return endpoints.images
        case .embeddings: return endpoints.embeddings
        case .files: return endpoints.files
        }
    }

    private static func validatedOfficialMetadataBaseURL(
        _ baseURL: String?,
        providerKind: ProviderKind
    ) -> String? {
        guard let baseURL else { return nil }
        guard providerKind != .relay, true else { return baseURL }
        guard let components = URLComponents(string: baseURL),
              components.scheme?.lowercased() == "https",
              let host = components.host?.lowercased(),
              allowedMetadataBaseURLDomains(for: providerKind).contains(where: { allowed in
                  host == allowed || host.hasSuffix(".\(allowed)")
              }) else {
            return nil
        }
        return baseURL
    }

    private static func allowedMetadataBaseURLDomains(for kind: ProviderKind) -> Set<String> {
        switch kind {
        case .openAI: return ["openai.com"]
        case .anthropic: return ["anthropic.com"]
        case .gemini: return ["googleapis.com"]
        case .openRouter: return ["openrouter.ai"]
        case .deepseek: return ["deepseek.com"]
        case .grok: return ["x.ai"]
        case .groq: return ["groq.com"]
        case .together: return ["together.xyz"]
        case .fireworks: return ["fireworks.ai"]
        case .miniMax: return ["minimax.io"]
        case .zhipu: return ["bigmodel.cn"]
        case .qwen: return ["aliyuncs.com"]
        case .moonshot: return ["moonshot.ai", "moonshot.cn"]
        case .mistral: return ["mistral.ai"]
        case .siliconFlow: return ["siliconflow.cn", "siliconflow.com"]
        case .relay: return []
        }
    }

    // MARK: - Built-in fallbacks

    static func fallbackBaseURL(for kind: ProviderKind) -> String {
        switch kind {
        case .openAI: return "https://api.openai.com/v1"
        case .anthropic: return "https://api.anthropic.com/v1"
        case .gemini: return "https://generativelanguage.googleapis.com/v1beta"
        case .openRouter: return "https://openrouter.ai/api/v1"
        case .deepseek: return "https://api.deepseek.com/v1"
        case .grok: return "https://api.x.ai/v1"
        case .groq: return "https://api.groq.com/openai/v1"
        case .together: return "https://api.together.xyz/v1"
        case .fireworks: return "https://api.fireworks.ai/inference/v1"
        case .miniMax: return "https://api.minimax.io/v1"
        case .zhipu: return "https://open.bigmodel.cn/api/paas/v4"
        case .qwen:
            // DashScope is region-split between `dashscope.aliyuncs.com` and
            // `dashscope-intl.aliyuncs.com`. The international host is the default; accounts
            // registered in the other region point at it with their own base URL.
            return "https://dashscope-intl.aliyuncs.com/compatible-mode/v1"
        case .moonshot: return "https://api.moonshot.ai/v1"
        case .mistral: return "https://api.mistral.ai/v1"
        case .siliconFlow: return "https://api.siliconflow.cn/v1"
        case .relay:
            return ""
        }
    }

    static func fallbackEndpointPath(_ providerKind: ProviderKind, kind: EndpointKind) -> String {
        switch (providerKind, kind) {
        case (.anthropic, .chat): return "/messages"
        case (.anthropic, .files): return "/files"
        case (.gemini, .chat), (.gemini, .responses):
            // Gemini encodes both the model id and the method in the URL
            // (`/models/<model>:generateContent`, or `:streamGenerateContent`), so the path
            // stops at `/models` and the caller appends the rest.
            return "/models"
        case (.gemini, .files): return "/files"
        case (.openAI, .chat): return "/chat/completions"
        case (.openAI, .responses): return "/responses"
        case (.openAI, .images): return "/images/generations"
        case (.openAI, .files): return "/files"
        case (.openRouter, .chat): return "/chat/completions"
        case (.qwen, .chat):
            return "/chat/completions"
        case (.qwen, .images):
            return "/api/v1/services/aigc/multimodal-generation/generation"
        case (.miniMax, .images):
            return "/image_generation"
        case (_, .chat): return "/chat/completions"
        case (_, .responses): return "/responses"
        case (_, .images): return "/images/generations"
        case (_, .embeddings): return "/embeddings"
        case (_, .files): return "/files"
        }
    }
}
